### PR TITLE
Generalized MultitaskGaussianLikelihood to include off-diagonal covariances for task noise

### PR DESCRIPTION
**Pull request by rajkumarkarthik**
Updates:

- Changed parameter of MultitaskGaussianLikelihood to include a rank = `rank` noise covariance matrix, parameterized by its covariance factor `task_noise_covar_factor`. 
- The full covariance matrix is the `v v' + e`, where `v` is `task_noise_covar_factor` and `e` is the diagonal matrix whose elements are `log_noise.exp()`. This is computed in `_eval_covar_matrix`.

The purpose of these updates is to make the cross-task noise covariance matrix general, and not just diagonal.

### CODE REVIEW

File: gpytorch/likelihoods/multitask_gaussian_likelihood.py [hunk 1]
```diff
@@ -1,41 +1,76 @@
...
-    exact inference.
+    for a full cross-task covariance structure for the noise. The fitted covariance matrix has rank `rank`.
+    If a strictly diagonal task noise covariance matrix is desired, then rank=0 should be set. (This option still
+    allows for a different `log_noise` parameter for each task.)
+
+    Like the Gaussian likelihood, this object can be used with exact inference.
     """
 
-    def __init__(self, n_tasks, log_task_noises_prior=None):
+    def __init__(self, n_tasks, rank=1, task_prior=None):
```
[the remark this answers is not in the document]
Comment (rajkumarkarthik) [reply gpytorch/likelihoods/multitask_gaussian_likelihood.py]: Done!

File: gpytorch/likelihoods/multitask_gaussian_likelihood.py [hunk 2]
```diff
@@ -51,12 +86,19 @@ def forward(self, input):
...
-        eye_lv = DiagLazyVariable(torch.ones(covar.size(-1) // self.n_tasks, device=self.log_noise.device))
-        task_var_lv = DiagLazyVariable(self.log_task_noises.exp())
+        eye_lv = DiagLazyVariable(
+            torch.ones(covar.size(-1) // self.n_tasks, device=self.log_noise.device)
+        )
+        if hasattr(self, "log_task_noises"):
+            task_var_lv = DiagLazyVariable(self.log_task_noises.exp())
+        else:
+            task_var_lv = RootLazyVariable(self.task_noise_covar_factor)
         diag_kron_lv = KroneckerProductLazyVariable(task_var_lv, eye_lv)
```
[the remark this answers is not in the document]
Comment (rajkumarkarthik) [reply gpytorch/likelihoods/multitask_gaussian_likelihood.py]: Changed to `covar_kron_lv`.

[the remark this answers is not in the document]
Comment (rajkumarkarthik) [reply gpytorch/likelihoods/multitask_gaussian_likelihood.py]: I think the way we are doing it is the right way. In the `rank = 0` case, we fit `D_t` that is not the identity matrix (or a scaled identity). Since we are just generalizing `D_t` when `rank > 0` we would preserve the order of operations.

**Comment by rajkumarkarthik** [comment]
Added unit test in new `likelihoods` folder in `gpytorch/tests`. Need to verify it actually ran, but fixed all lint and other issues.

**Comment by jacobrgardner** [comment]
Looks good to me!

**Comment by Balandat** [comment]
So looks like this breaks the `test_multitask_gp_mean_abs_error` unit test - not sure why exactly, but  the` likelihood.task_noise_covar_factor` stays at zero during the fitting. So either there is something wrong with the gradients, or the test condition is wrong, or this is some kind of identifiability issue.

**Comment by jacobrgardner** [comment]
I believe the test in question doesn't use the covar factor because `root=0` by default:
https://github.com/cornellius-gp/gpytorch/blob/0c46cb2a2e7f93d84708fdf2a09421537085b7e4/test/examples/test_kronecker_multitask_gp_regression.py#L58

@rajkumarkarthik included a unit test that verified those parameters got trained when `rank>0`. The puzzling thing is why that parameter even exists in that test?

**Comment by jacobrgardner** [comment]
Oh I see, the new test slipped by with the same name :-(
